src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp

```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
-
-    // First bound test.
-    if (upperBounds(i) <= m)
+    arma::mat threadNewCentroids(centroids.n_rows, centroids.n_cols,
+      arma::fill::zeros);
+    arma::Col<size_t> threadCounts(centroids.n_cols, arma::fill::zeros);
+    size_t threadHamerlyPruned = 0;
+    size_t threadDistanceCalculations = 0;
+
+    #pragma omp for schedule(static)
+    for (size_t i = 0; i < dataset.n_cols; ++i)
     {
-      ++hamerlyPruned;
-      newCentroids.col(assignments[i]) += dataset.col(i);
-      ++counts(assignments[i]);
-      continue;
-    }
+      const double m = std::max(minClusterDistances(assignments[i]),
+                                lowerBounds(i));
 
-    // Tighten upper bound.
-    upperBounds(i) = distance.Evaluate(dataset.col(i),
-                                       centroids.col(assignments[i]));
-    ++distanceCalculations;
-
-    // Second bound test.
-    if (upperBounds(i) <= m)
-    {
-      newCentroids.col(assignments[i]) += dataset.col(i);
-      ++counts(assignments[i]);
-      continue;
-    }
-
-    // The bounds failed.  So test against all other clusters.
-    // This is Hamerly's Point-All-Ctrs() function from the paper.
-    // We have to reset the lower bound first.
-    lowerBounds(i) = DBL_MAX;
-    for (size_t c = 0; c < centroids.n_cols; ++c)
-    {
-      if (c == assignments[i])
+      // First bound test.
+      if (upperBounds(i) <= m)
+      {
+        ++threadHamerlyPruned;
+        threadNewCentroids.col(assignments[i]) += dataset.col(i);
+        ++threadCounts(assignments[i]);
         continue;
+      }
 
-      const double dist = distance.Evaluate(dataset.col(i), centroids.col(c));
+      // Tighten upper bound.
+      upperBounds(i) = distance.Evaluate(dataset.col(i),
+                                         centroids.col(assignments[i]));
+      ++threadDistanceCalculations;
 
-      // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i)).
-      if (dist < upperBounds(i))
+      // Second bound test.
+      if (upperBounds(i) <= m)
       {
-        // lowerBounds holds the second closest cluster.
-        lowerBounds(i) = upperBounds(i);
-        upperBounds(i) = dist;
-        assignments[i] = c;
+        threadNewCentroids.col(assignments[i]) += dataset.col(i);
+        ++threadCounts(assignments[i]);
+        continue;
       }
-      else if (dist < lowerBounds(i))
+
+      // The bounds failed.  So test against all other clusters.
+      // This is Hamerly's Point-All-Ctrs() function from the paper.
+      // We have to reset the lower bound first.
+      lowerBounds(i) = DBL_MAX;
+      for (size_t c = 0; c < centroids.n_cols; ++c)
       {
-        // This is a closer second-closest cluster.
-        lowerBounds(i) = dist;
+        if (c == assignments[i])
+          continue;
+
+        const double dist = distance.Evaluate(dataset.col(i), centroids.col(c));
+
+        // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i))
```

Comment:
```suggestion
        // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i)).
```
The original period seems to have gotten lost 😄

Reply:
i removed it on purpose because it is above the 80 char limit, due to the indent, it is the 81st char. i thought removing it would be better than breaking the line and causing a larger diff size. let me know what is best here :)